prepare v1.0.0 release

rursprung opened the pull request:
see the individual commit messages for more details.

the only "violation" of the general guidelines i could see was the (optional) dependency to defmt v0.3 (which many claim is de-facto 1.0). with the dependency renamed to `defmt-03` it should be possible to subsequently add a `defmt-10` feature to support both at the same time.

the other open point is implementing `std::error::Error` for the error enums. that can come once rust 1.81.0 has been released which moves this to `core::error:Error`, so no `std` feature will be required here (as per our README we can bump the MSRV w/o a major release, thus it's ok to bump it to 1.81.0 once that has been released).

thus, once this PR has been reviewed & merged it should be fine to release v1.0.0 and close #4.

@reviewers: if possible, please give the whole crate a glance again to make sure that we're not missing anything relevant for v1.0.0. thanks!

eldruin commented on CHANGELOG.md:
```diff
@@ -15,9 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
-* Renamed `DriveCommand::Backwards` to `DriveCommand::Backward` to match
+* **Breaking**: update to `embedded-hal` 1.0
+* **Breaking**: Renamed error types to their struct names
+* **Breaking**: Renamed `DriveCommand::Backwards` to `DriveCommand::Backward` to match
   `DriveCommand::Forward`
```
```suggestion
* **Breaking**: Renamed `DriveCommand::Backwards` to `DriveCommand::Backward` to match
  `DriveCommand::Forward`.
* **Breaking**: Renamed `defmt` feature `defmt-03`.
```

eldruin commented on README.md:
```diff
@@ -21,20 +21,27 @@ See the documentation for usage examples.
+## License
+Licensed under either of
+
+* Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
```
```suggestion
* Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or <http://www.apache.org/licenses/LICENSE-2.0>)
```

rursprung replied:
i've changed these, though i have copied this verbatim from [the API guidelines](https://rust-lang.github.io/api-guidelines/necessities.html#crate-and-its-dependencies-have-a-permissive-license-c-permissive). maybe this should be changed there as well?

eldruin commented on README.md:
```diff
@@ -21,20 +21,27 @@ See the documentation for usage examples.
+Licensed under either of
+
+* Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
+* MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)
```
```suggestion
* MIT license ([LICENSE-MIT](LICENSE-MIT) or <http://opensource.org/licenses/MIT>)
```

eldruin reviewed (commented):
Thank you for your work!
This PR seems fine to me but I have not had a look at the whole source code.
I would suggest having a look at the checklist [here](https://rust-lang.github.io/api-guidelines/checklist.html).
One thing I noticed is that the dependency to `embedded-hal-mock` can be customized so that this does not depend on `embedded-hal` 0.2 with:
```toml
embedded-hal-mock = {version = "0.11", default-features=false, features=["eh1"]}
```

ryan-summers reviewed (commented):
The rest here is a general review of the codebase as it stands. Feel free to ignore any of these if you disagree:
* I suspect you could simplify your error definitions by defining a single `embedded_hal::digital::ErrorType` for all of the pins and assume they all have the same error type. It would be _exceptionally weird_ if they didn't. And even in the edge case where i.e. some pins are on a GPIO expander and others are on the MCU, a user could make their own error enum to wrap them up into a single type expression. This might aid in usability of the driver.
* Same for the above with PWM errors
* `DriveCommand` specifies a speed in percentage, but the stored value is a u8. Are they supposed to use 0 for 0% and u8::MAX for 100%?
* Beware exposing `pub` fields in `Tb6612fng`. Read through https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private for an analysis of why. It recommends exposing the fields as getter methods.
* As it stands, your constructor just takes a whole bunch of pins. This could make it easy to get them mixed up. Maybe a builder pattern would be better to ensure the pins are going to their proper definitions? Or have the `Tb6612fng take in two `Motors` instead.
*

ripytide commented:
> I suspect you could simplify your error definitions by defining a single embedded_hal::digital::ErrorType for all of the pins and assume they all have the same error type. It would be exceptionally weird if they didn't. And even in the edge case where i.e. some pins are on a GPIO expander and others are on the MCU, a user could make their own error enum to wrap them up into a single type expression. This might aid in usability of the driver.
Same for the above with PWM errors

I'm not sure I understand this suggestion, are you suggesting we create a catch-all error type? Such as:

```rust
struct Error;
impl Tb6612fng {
    fn drive(..) -> Result<(), Error>;
}
```

This has already been discussed as being less flexible due to the scenario you described as having pins with different error types. How would a user create their own wrapping error types though? Would that not require a generic return type? Such as:

```rust
impl Tb6612fng {
    fn drive<E: embedded_hal::digital::ErrorType>(..) -> Result<(), E>;
}
```

This seems harder to use since it adds trait bounds to the `drive()` function which can cause hard to diagnose trait bound compile errors.

> DriveCommand specifies a speed in percentage, but the stored value is a u8. Are they supposed to use 0 for 0% and u8::MAX for 100%?

This is since we pass the `u8` to `embedded_hal::SetDutyCycle::set_duty_cycle_percent(speed: u8)`, this could be better documented on our side though about 0-100 being required. Possibly, we could also switch to a `DrivePercent` and `DriveFraction` enum so a user could pick which method they wanted to use (`SetDutyCycle::set_duty_cycle_percent()` being the other possible method).

> Beware exposing pub fields in Tb6612fng. Read through https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private for an analysis of why. It recommends exposing the fields as getter methods.

Switching to getting mutable motor references via getters also has some drawbacks such as preventing multiple mutable references otherwise allowed:
```rust
let driver = Tbb6612fng::new(...);

//no borrow checker issue
takes_mutable_reference(&mut driver.motor_a);
takes_mutable_reference(&mut driver.motor_b);

//borrow checker issue
takes_mutable_reference(driver.motor_a_mut());
takes_mutable_reference(driver.motor_b_mut());
```

> As it stands, your constructor just takes a whole bunch of pins. This could make it easy to get them mixed up. Maybe a builder pattern would be better to ensure the pins are going to their proper definitions? Or have the Tb6612fng take in two Motors` instead.

Good point, `Tb6612fng::new(motor_a: Motor, motor_b: Motor, stby: STBY)` sounds much nicer to use.

rursprung commented on Cargo.toml:
```diff
@@ -9,6 +9,10 @@ repository = "https://github.com/rust-embedded-community/tb6612fng-rs"
+authors = ["Ralph Ursprung <ralph.ursprung@gmail.com>", "ripytide <james.forsterer@gmail.com>"]
+
+[features]
+defmt-03 = ["dep:defmt"]
```
i've now removed the `defmt` dependency. you're right, it's not needed - and consumers don't even need to wrap it in a custom type, they can just use `defmt::Debug2Format` as a call:
```rust
defmt::info!("{}", defmt::Debug2Format(&command));
```

rursprung commented:
thanks a lot for your review! sorry for not reacting on it earlier, but i was away!


> I would suggest having a look at the checklist [here](https://rust-lang.github.io/api-guidelines/checklist.html).

i had already done that, thanks!

> One thing I noticed is that the dependency to `embedded-hal-mock` can be customized so that this does not depend on `embedded-hal` 0.2 with:
> `embedded-hal-mock = {version = "0.11", default-features=false, features=["eh1"]}

thanks! i've changed it now and also raised https://github.com/dbrgn/embedded-hal-mock/pull/122

> * I suspect you could simplify your error definitions by defining a single `embedded_hal::digital::ErrorType` for all of the pins and assume they all have the same error type. It would be _exceptionally weird_ if they didn't. And even in the edge case where i.e. some pins are on a GPIO expander and others are on the MCU, a user could make their own error enum to wrap them up into a single type expression. This might aid in usability of the driver.
> * Same for the above with PWM errors

as per @ripytide's comment i don't thinkt hat this simplification would actually work (or simplify anything)?


> * `DriveCommand` specifies a speed in percentage, but the stored value is a u8. Are they supposed to use 0 for 0% and u8::MAX for 100%?

the comment on the enum says that it's "in percentage" - isn't that enough to show that it's in the 0..100 range? (since it's `u8` it can't be 0..1, so that confusion at least is avoided). pulling in something like [bounded-integer](https://crates.io/crates/bounded-integer) (or hand-rolling it) seems like an overkill here?

> * Beware exposing `pub` fields in `Tb6612fng`. Read through https://rust-lang.github.io/api-guidelines/future-proofing.html#structs-have-private-fields-c-struct-private for an analysis of why. It recommends exposing the fields as getter methods.

i am aware that it's not recommended to have `pub` fields, but as pointed out by @ripytide it makes it so much more convenient to use it since you need `mut` access to the `Motor` to use it.

> * As it stands, your constructor just takes a whole bunch of pins. This could make it easy to get them mixed up. Maybe a builder pattern would be better to ensure the pins are going to their proper definitions? Or have the `Tb6612fng take in two `Motors` instead.

thanks, you're right! i've changed this now

eldruin commented on CHANGELOG.md:
```diff
@@ -7,24 +7,30 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+* **Breaking**: Renamed error types to their struct names
+* **Breaking**: Renamed `DriveCommand::Backwards` to `DriveCommand::Backward` to match
   `DriveCommand::Forward`
+* The MSRV has been updated to 1.81.0 due to `core::error::Error being implemented`
```
```suggestion
* The MSRV has been updated to 1.81.0 due to `core::error::Error` being stabilized.
```

rursprung replied:
🙈

eldruin reviewed (commented):
From a quick look the change seems fine to me but I have not checked the whole code. Releasing a 2.0 for this crate should not be a big problem, though.

eldruin reviewed (approved):
Thank you!
